### PR TITLE
Basic auth flow

### PR DESCRIPTION
**Pull request by msujew**
Implements a basic auth flow. Shows a login page when the server has been started with `--password=<password>`. Inserting the correct password redirects to the application, while a wrong password brings the user back to the login.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,8 +59,11 @@
   },
   "dependencies": {
     "@vscode/vscode-languagedetection": "1.0.15",
+    "argon2": "^0.28.2",
     "applicationinsights": "1.0.8",
     "chokidar": "3.5.1",
+    "cookie-parser": "^1.4.5",
+    "express": "^4.17.1",
```
Comment (akosyakov) [inline package.json:66]: Is there a way to build auth on plain http without pulling express?

Comment (akosyakov) [reply package.json:66]: Although let's go with it if it is easier. We need to make sure that it still works in budnled variant.

Do follwoing: 
- yarn gulp server
- it should product the prod ready server under `/workspace`
- then try to start it and make sure that deps are found, if not then something should be changed in `/workspace/vscode/remote` folder.

Comment (akosyakov) [reply package.json:66]: Generally adding new deps is not welcomed.

Comment (msujew) [reply package.json:66]: It's generally possible (see every commit before 61bc66c). But that relies on manually parsing cookies and request bodies, which works, but looks really hacky. I don't care, the express solution is just way cleaner.

I can remove all new dependencies if you want. We can keep it at plain passwords without the `argon2` dependency as well.

Comment (akosyakov) [reply package.json:66]: try to build it for final distribution and run if you make it work let's go with it. We can revert later, we don't promise backward compatibility here to anyone :D

Comment (msujew) [reply package.json:66]: Which file/script starts the production server? Running `server.js` under `server-pkg/out` does nothing.

Comment (akosyakov) [reply package.json:66]: correct

**Comment by akosyakov** [comment]
One more thing basic oauth should not be contributed to server.main.ts, only via server.ts. We are not going to use it in Gitpod and should not have unnecesray deps then.

File: package.json (hunk 1 above)
Comment (msujew) [reply package.json:66]: > correct

Is it supposed to do nothing? If I start it, I end up back at the console without any message.

**Comment by akosyakov** [comment]
Can we revert to minimal change? I cannot really understand what was changed in server.main.ts. 

The PR should be:
- introduce a needed callback in server.main.ts
- implement rest in server.ts
- add login page

There is no need for unnecessary refactoring and so on.

**Comment by akosyakov** [comment]
Web sockets are not secured.

File: src/vs/server/node/auth-http.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		if (pathname === '/login') {
+			const password = (await collectRequestData(req)).password;
+			const { isPasswordValid, hashedPassword } = await handlePasswordValidation({
+				passwordFromRequestBody: password,
+				passwordFromArgs: args.password,
+			});
+
+			if (isPasswordValid) {
+				res.writeHead(302, {
```
Comment (akosyakov) [inline src/vs/server/node/auth-http.ts:33]: can we do without a redirect?

I would imagine that we add a new callback like `verifyClient` which is used for all incoming requests and returns boolean value whether we need to continue or not. We don't need to reuse `handleRequest` for it. This function will be called in ws upgrade callback as well as in http listener.

File: src/vs/server/node/auth-http.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+					'Location': '/',
+					'Set-Cookie': `key=${hashedPassword}`,
```
Comment (akosyakov) [inline src/vs/server/node/auth-http.ts:35]: please add `HttpOnly` attribute

I'm not actually sure that auth with cookies work like that, i.e. caching the password in the client browser. Usually password is send only once and then session cookie is generated which is stored in server and set as a cookie to cache in the browser?

File: src/vs/server/node/server.main.ts [hunk 4]
```diff
@@ -571,10 +562,18 @@ export async function main(options: IServerOptions): Promise<void> {
...
+				if (options.verifyRequest) {
+					verified = await instantiationService.invokeFunction(accessor => options.verifyRequest!(req, accessor));
+				}
+				const verifiedReq = req as VerifiedIncomingMessage;
+				verifiedReq.verified = verified;
+				if (options.handleRequest && await instantiationService.invokeFunction(accessor => options.handleRequest!(pathname, verifiedReq, res, accessor, channelServer))) {
 					return;
 				}
-
+				if (!verifiedReq.verified) {
```
Comment (akosyakov) [inline src/vs/server/node/server.main.ts:574]: Can it be just:
```ts
if (options.verifyRequest && !(await instantiationService.invokeFunction(accessor => options.verifyRequest(req, accessor))) {
    return serveError(req, res, 401, 'Unauthorized');
}
```

Comment (msujew) [reply src/vs/server/node/server.main.ts:574]: Well, not really (I would have liked to implement it that way). We still need to send the login page back, if we haven't verified the client yet. Also, we need to change the flag, if we verify during the request (like receiving a form data request with the password)

File: src/vs/server/node/server.main.ts [hunk 5]
```diff
@@ -899,8 +902,8 @@ export async function main(options: IServerOptions): Promise<void> {
 			});
 		});
 		let port = 3000;
-		if (parsedArgs.port) {
-			port = Number(parsedArgs.port);
+		if (args.port) {
```
[the remark this answers is not in the document]
Comment (msujew) [reply src/vs/server/node/server.main.ts:905]: The `parsedArgs` where only available locally inside of the `server.main` method. I needed them outside of that as well, to have access to the password.

File: src/vs/server/node/auth-http.ts (hunk 2 above)
Comment (msujew) [reply src/vs/server/node/auth-http.ts]: I added a new verification handler and did it without the redirect. The only caveat is that when users login and try to reload the app, the browser will ask to resend form information.

File: src/vs/server/node/server.main.ts (hunk 4 above)
Comment (akosyakov) [reply src/vs/server/node/server.main.ts:574]: Can we hide then `serveError(req, res, 401, 'Unauthorized')` behind `verifyRequest`? i.e.:
```ts
if (options.verifyRequest && !(await instantiationService.invokeFunction(accessor => options.verifyRequest(req, accessor))) {
    return;
}
```

Comment (msujew) [reply src/vs/server/node/server.main.ts]: Alright, done 👍

**Comment by akosyakov** [comment]
@msujew Could you rebase please? Some minor comments:
- for login screen can we center it, i.e like on google.com and add a comment about the password in server terminal?
- there is no way to fix pop up on first reload about the form fields, do you know the reason?

**Comment by msujew** [comment]
> there is no way to fix pop up on first reload about the form fields, do you know the reason?

Well, the alternative would be the redirect. I don't think there's a way around that, as that's something implemented by the browser, not by us.

**Comment by akosyakov** [comment]
We discussed it and decided to put it on hold, try without built-in auth for now. Please don't be discourage, contribution is good!

**Comment by rob-64** [comment]
I was able to accomplish this with a Caddy reverse proxy in front and basic auth there.

**Comment by akosyakov** [comment]
I'm closing it. The goal of the project just to provide easy to maintain open source server. Integration issues should be resolved by adopters or end users.

That said, it does not mean that there could not be guides showing how to securely setup the server. Anyone is welcomed to contribute them.

**Pull request closed without merging**